From 237997cb170af9fd3e31c6f054b186d6b3c064a4 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:53 -0400
Subject: [ethernet] i40e: Fix a boundary condition and turning off of ntuple

Message-id: <1409658042-22689-233-git-send-email-sassmann@redhat.com>
Patchwork-id: 90434
O-Subject: [RHEL7.1 PATCH 232/281] i40e: Fix a boundary condition and turning off of ntuple
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

When turning off ntuple with a FD table full situation,
the driver would have auto disabled FD filter additions.
Clear the auto disable flag for FD_SB so that when the
feature is turned on again using "ethtool -K ethx ntuple on"
we can start adding filters once again.

Change-ID: I036a32e7331bcae765b657c8abb4fa070940b163
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 8a4f34fbef029771f686ee93311fb2b488247b16)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 529ab3e..36e9797 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -6885,9 +6885,11 @@ bool i40e_set_ntuple(struct i40e_pf *pf, netdev_features_t features)
    i40e_fdir_filter_exit(pf);
   }
   pf->flags &= ~I40E_FLAG_FD_SB_ENABLED;
-  /* if ATR was disabled it can be re-enabled. */
-  if (!(pf->flags & I40E_FLAG_FD_ATR_ENABLED))
-   pf->flags |= I40E_FLAG_FD_ATR_ENABLED;
+  pf->auto_disable_flags &= ~I40E_FLAG_FD_SB_ENABLED;
+  /* if ATR was auto disabled it can be re-enabled. */
+  if ((pf->flags & I40E_FLAG_FD_ATR_ENABLED) &&
+      (pf->auto_disable_flags & I40E_FLAG_FD_ATR_ENABLED))
+   pf->auto_disable_flags &= ~I40E_FLAG_FD_ATR_ENABLED;
  }
  return need_reset;
 }
-- 
1.7.1